openhtf/__init__.py

```diff
@@ -159,21 +164,27 @@ def add_output_callbacks(self, *callbacks):
     self._test_options.output_callbacks.extend(callbacks)
 
   def configure(self, **kwargs):
-    """Update test-wide configuration options.
-
-    Valid kwargs:
-      output_callbacks: List of output callbacks to run, typically it's better
-          to use add_output_callbacks(), but you can pass [] here to reset them.
-      teardown_function: Function to run at teardown.  We pass the same
-          arguments to it as a phase.
-    """
+    """Update test-wide configuration options. See TestOptions for docs."""
     # These internally ensure they are safe to call multiple times with no weird
     # side effects.
     create_arg_parser(add_help=True).parse_known_args()
     logs.setup_logger()
     for key, value in kwargs.iteritems():
       setattr(self._test_options, key, value)
 
+    # 'capture_source' takes effect here.
+    if self._test_options.capture_source:
+      # First, we copy the phases with the real CodeInfo for them.
+      phases = [
+        mutablerecords.CopyRecord(
+            phase, code_info=test_record.CodeInfo.for_function(phase.func))
+        for phase in self._test_desc.phases]
+
+      # Then we replace the TestDescriptor with one that stores the test
+      # module's CodeInfo as well as our newly copied phases.
+      self._test_desc = self._test_desc._replace(
+          code_info=self.code_info, phases=phases)
```

Comment:
I have two qualms with using `_replace` on the descriptor as done here. First, I think it would be better style-wise to explicitly export `_replace` as `replace` on the `TestDescriptor` class.

Second, and perhaps more concerning, is that any references to `self._test_desc` that are grabbed or given out before this `_replace` will be defunct after it. Right now anyone can grab it through the `descriptor` property, and there's nothing to stop someone grabbing the original one after instantiation, but before `configure()` is called. One thought is to maybe use a mutable record here instead, but I think we specifically chose not to because the `TestDescriptor` is intended to be static between runs of the same test.